### PR TITLE
feat: semver implementation

### PR DESCRIPTION
**Pull request by ozayr-zaviar**
## Summary
- Semantic version added
- Greater than or equal (ge) and less than or equal (le) evaluator added

## Test plan
- All relevant test cases added

### CODE REVIEW

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+const logger = getLogger();
+
+/**
+  * Evaluate if provided string is number only
+  * @param  {unknown}  content
+  * @return {boolean}  true if the string is number only
+  *                     
+  */
+  function isNumber(content: string): boolean {
+    return new RegExp(/^\d+$/).test(content);
```
Comment (msohailhussain) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:29]: remove `RegExp`
`/^\d+$/.test(content)`

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  }
+  
+  /**
+   * check if there is any white spaces " " in version
+   * @param  {unknown}  version
+   * @return {boolean}  true if the version contains " "
+   *                     
+   */
+  function hasWhiteSpaces(version: string): boolean {
+    return new RegExp(/\s/).test(version);
```
Comment (msohailhussain) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:81]: `/\s/.test(version)`

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+  }
+  
+  /**
+   * split version in parts
+   * @param  {unknown}  version
+   * @return {boolean}  The array of version split into smaller parts i.e major, minor, patch etc
+   *                    null if given version is in invalid format 
+   */
+  function splitVersion(version: string): string[] | null {
```
Comment (msohailhussain) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:90]: https://github.com/optimizely/javascript-sdk/pull/566#discussion_r489739380

Comment (msohailhussain) [reply packages/optimizely-sdk/lib/utils/semantic_version/index.ts:90]: @mjc1283 I think, parsing condition would be bit difficult if we return SemVer interface like you said in the comment.
One of the reason, we are allowing all variations. Version can have the following forms. 
major
major.minor
major.minor.patch
major.minor.patch.pre-release
major.minor.patch.pre-release+build
major.minor.patch+build

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+   * Compare user version with condition version
+   * @param  {string}  conditionsVersion
+   * @param  {string}  userProvidedVersion
+   * @return {number | null}  0 if user version is equal to condition version
+   *                          1 if user version is greater than condition version
+   *                         -1 if user version is less than condition version
+   *                          null if invalid user or condition version is provided
+   */
+  export function compareVersion(conditionsVersion: string, userProvidedVersion: string): number | null {
+    const isPreReleaseInconditionsVersion = isPreReleaseVersion(conditionsVersion);
```
Comment (msohailhussain) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:151]: I think we don't really need this variable, let me think. btw
`x isPreReleaseInConditionsVersion`

File: packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.ts [hunk 5]
```diff
@@ -165,16 +187,13 @@ function existsEvaluator(condition: Condition, userAttributes: UserAttributes):
...
- * @param   {LoggerFacade}    logger
- * @returns {?boolean}        true if the user attribute value is greater than the condition value,
- *                            false if the user attribute value is less than or equal to the condition value,
- *                            null if the condition value isn't a number or the user attribute value
- *                            isn't a number
+ * @returns {?boolean}        true if values are valid,
+ *                            false if values are not valid
  */
-function greaterThanEvaluator(condition: Condition, userAttributes: UserAttributes): boolean | null {
+function validateValues(condition: Condition, userAttributes: UserAttributes): boolean {
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/core/custom_attribute_condition_evaluator/index.ts:196]:
```suggestion
function validateValuesForNumericCondition(condition: Condition, userAttributes: UserAttributes): boolean {
```

File: packages/optimizely-sdk/lib/utils/semantic_version/index.ts [hunk 6]
```diff
@@ -0,0 +1,190 @@
+/**
```
Comment (mjc1283) [inline packages/optimizely-sdk/lib/utils/semantic_version/index.ts:1]: This file has several style/formatting issues. Please see other files and follow the same style for indentation, brackets on the same line as `if` statements, always use brackets with `if` statements. I will try update the linter to enforce these.

**Review by mjc1283 — CHANGES_REQUESTED** [review]

**Review by mjc1283 — APPROVED** [review]

**Pull request merged**